Fix clicking in the TOC does not scroll

parmentelat opened the pull request:
workaround for cases where headingToElement map is not properly filled

<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

fixes #14591  
intended to supersede #14884

<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes

as suggested by @fcollonval in this post
https://github.com/jupyterlab/jupyterlab/pull/14884#pullrequestreview-1568943032

the `onActiveHeadingChanged` function (notebook viewer)
uses a fallback strategy in pathological cases where `headingToElement` map is not properly filled

<!-- Describe the code changes and how they address the issue. -->

## User-facing changes

no visual change; clicking in a TOC entry should more reliably scroll to the corresponding cell

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

none

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

parmentelat commented:
there are several possible further directions to this PR

- @fcollonval also suggested, in case the target cell is in edit mode, to select the title  
  imho this is not necessarily a good idea, mentioned here for the record
- I have observed similar flaws in the `markdownviewer` area; however the approach used here is not applicable there

regardless, in my experience this first step would already be of great help, particularly in the context of
https://github.com/executablebooks/jupyterlab-myst/issues/184

also, I went for starting on top of the `4.0.x` branch, because that's the version I would be targetting, be should it have gone onto `main` instead ?

krassowski commented:
Yes, it should be targeting the `main` branch.

parmentelat commented:
hi @krassowski 
please let me know if there's anything I can do to help this ship as early as possible
I am massively using jupyterlab-myst and this issue basically ruins navigation through the TOC, which is a major productivity impediment, so again I'm highly motivated here ;)

krassowski commented:
Adding a test case which demonstrates the issue on `main` and fixes it here would help.

parmentelat commented:
starting here
https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#visual-regression-and-ui-tests

I went to there
https://github.com/jupyterlab/jupyterlab/blob/main/galata/README.md

where I was instructed to do this:

```
cd jupyterlab-repo-top

jlpm add -D @jupyterlab/galata
→ ... ➤ YN0000: Done with warnings in 2s 676ms

# so far so good, but

jlpm playwright install
→ Usage Error: Couldn't find a script named "playwright".
```

what am I doing wrong ?

krassowski commented:
The https://github.com/jupyterlab/jupyterlab/blob/main/galata/README.md is a bit odd because it is a mix of instructions for use with jupyterlab and with extensions and it has a tendency to get outdated. There is no `jlpm playwright` when running galata in jupyterlab, what I use is:

```bash
jlpm install # just once
# then check nothing is running on port 8888
jlpm run start
# then wait for jupyterlab to start, open another terminal
jlpm run test --project jupyterlab --grep "testofinterest"
```

If a lab instance is running on 8888 and stuck (i.e. you cannot shut it down) I use `kill -9 $(sudo lsof -i -P -n | grep LISTEN | grep 8888 | awk '{ print $2 }')`

parmentelat commented:
thanks for the tips
before I pour my own test in the mix, I'm trying to run the 'official' test named `toc.test.ts`
but there's clearly something fishy with the `test-toc` temporary folder, that I don't have on my filesystem, and that causes the test to fail with errors like this 
```
[W 2023-10-01 16:31:22.915 ServerApp] 404 GET /api/contents/test-toc?type=directory&content=1 (::1): No such file or directory: test-toc
```

is there anythong I should know about `test-toc` ?
```

fcollonval commented:
Thanks @parmentelat 

The code looks good - if you could add the test in this PR, it will be easier to validate it and produce the _correct_ reference snapshot.

parmentelat commented:
please update galata snapshots

parmentelat commented:
once the linux snapshot is produced and it checks out as compared to the darwin one, we can remove the latter

fcollonval commented:
FYI the commit updating snapshots are not triggering the CI (some limitation of GitHub to avoid infinite job run). To force the CI to be executed the trick is to close and reopen the PR, don't hesitate to do it when you see the new snapshot commit.

parmentelat commented:
shaking the PR

parmentelat commented:
please update galata snapshots

parmentelat commented:
@fcollonval 
I have closed and reopened the PR, but am still struggling to get the snapshot updated
please advise...

krassowski commented on packages/notebook/src/toc.ts:
```diff
@@ -530,6 +530,9 @@ export class NotebookToCFactory extends TableOfContentsFactory<NotebookPanel> {
             }
+          } else {
+            console.debug('scrolling to heading: using fallback strategy');
+            widget.content.scrollToItem(widget.content.activeCellIndex);
```
`Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator`

fcollonval replied:
To provide some guide line to fix this. I suggest changing:
* `onCellInViewport` to be `async (cell: Cell): Promise<void>`
* adding `await in front of Line 535
* adding `return` at line 550 so that the catch line 552 will catch everything

krassowski reviewed (commented):
I see the galata update actions running on CI, one started two hours ago but was only picked up an hour ago: https://github.com/jupyterlab/jupyterlab/actions/runs/6381080519/job/17316801930 (probably will fail to push updates though if your forced pushed in the mean time) another is waiting (https://github.com/jupyterlab/jupyterlab/actions/runs/6382766999). There was a few merges, and a few authors rebased their PRs eagerly which lead to CI falling behind (in addition to being generally slower since last month).

parmentelat commented:
I am no gh actions expert as you have noticed, so thanks for having my back :)

the linux snapshot is now updated and so I have removed the darwin snapshot

parmentelat commented:
I am unsure how to proceed from this point on, is there anything else I need to do to get this merged ?

krassowski commented:
It looks the liner is still falling:

```
2023-10-05T07:04:07.4461474Z + [[ lint == integrity ]]
2023-10-05T07:04:07.4461789Z + [[ lint == lint ]]
2023-10-05T07:04:07.4583259Z + jlpm run prettier:check
2023-10-05T07:04:08.9120557Z Checking formatting...
2023-10-05T07:04:32.1575423Z [warn] packages/notebook/src/toc.ts
2023-10-05T07:04:41.4732063Z [warn] Code style issues found in the above file. Run Prettier to fix.
2023-10-05T07:04:41.5334848Z + echo 'Please run `jlpm run prettier` locally and push changes'
```

parmentelat commented:
thanks @krassowski I will take care of it
also I realize I did an unintentional merge, is it preferable that I rebase instead ?

krassowski commented:
No, is fine we do squash merges anyways.

krassowski commented:
I tested this locally and it fixes one part of the issue (scrolling on click when switching between headings), but does not seem to fix the problem of lack of scrolling when clicking the same header twice. @parmentelat can you confirm that you also see the same:
1. Click on a heading in the ToC
2. Scroll down
3. Click on the same heading again

Does it scroll back to the heading for you? Note: I am not saying that we should not merge it as-is (we should, it is a great improvement), just want to make sure that we have a clarity on whether merging this PR closes https://github.com/jupyterlab/jupyterlab/issues/14591 or not.

parmentelat commented:
I can confirm your observation, if I follow your scenario the clicked heading does not come back into view

somehow it does not surprise me because we're talking about a callback that's named `onActiveHeadingChanged`, and in your scenario the active heading, well, does not change...

I can see how this can feel off in terms of user experience, obviously

parmentelat commented:
is there a chance this may be merged for the next official release ? 
is there anything else I can do to this end ?

krassowski commented on galata/test/jupyterlab/toc-scrolling.test.ts:
```diff
@@ -0,0 +1,56 @@
+    await page.notebook.activate(fileName);
+
+    await page.sidebar.openTab('table-of-contents');
+
+    // await page.click('.jp-toc-numberingButton');
```
I guess this is not used anymore?

```suggestion
```

krassowski reviewed (approved):
One tiny suggestion, otherwise it looks good to me.

parmentelat commented:
for the record, the very same code - modulo the harmless removal of a comment line in the very last commit -  has been passing all the checks
I am at a loss to know what to do more on this one, please merge if everything is ok

krassowski commented:
No worries, some of the checks are just timing out/flaky (i restarted these)  and one (`
Linux JS Tests / JS (js-metapackage) `) is an issue which will be fixed by https://github.com/jupyterlab/jupyterlab/pull/15237 but that needs review before we can proceed with merging it and then other PRs.

krassowski commented:
I hope to merge this one (along a few others) and release a new point release this week.

fcollonval reviewed (approved):
Thanks @parmentelat

fcollonval commented:
@meeseeksdev please backport to 4.0.x
